File: Backtracking/permute.js
const permute = (nums) => {
  const backtrack = (current, used) => {
    if (current.length === nums.length) {
      permutations.push([...current]);
      return;
    }
    for (let num of nums) {
      if (!used.has(num)) {
        current.push(num);
        used.add(num);
        backtrack(current, used);
        current.pop();
        used.delete(num);
      }
    }
  };

  const permutations = [];
  backtrack([], new Set());
  return permutations;
};

// Time O(n⋅n!)
// Space O(n⋅n!)

const nums = [1, 2, 3];
console.log(permute(nums));

File: Backtracking/permute.py
from typing import List

class Permute:
  def permute(self, nums: List[int]) -> List[List[int]]:
  
    def backtrack(current: List[int], used: set[int]):
      if len(current) == len(nums):
        permutations.append([*current])
        return
      for num in nums:
        if num not in used:
          current.append(num)
          used.add(num)
          backtrack(current, used)
          current.pop()
          used.remove(num)

    permutations = []
    backtrack([], set())
    return permutations
  
# Time: O(n * n!) where n is the length of nums
# Space: O(n!) for the output list and O(n) for the recursion stack, total O(n + n!)